src/Microsoft.AspNetCore.SignalR.Core/Internal/HubMethodDescriptor.cs

```diff
@@ -0,0 +1,136 @@
+
+        public HubMethodDescriptor(ObjectMethodExecutor methodExecutor, IEnumerable<IAuthorizeData> policies)
+        {
+            MethodExecutor = methodExecutor;
+            ParameterTypes = methodExecutor.MethodParameters.Select(p => p.ParameterType).ToArray();
+            Policies = policies.ToArray();
+
+            NonAsyncReturnType = (MethodExecutor.IsMethodAsync)
+                ? MethodExecutor.AsyncResultType
+                : MethodExecutor.MethodReturnType;
+
+            if (IsObservableType(NonAsyncReturnType, out var observableItemType))
+            {
+                IsObservable = true;
+                StreamReturnType = observableItemType;
+            }
+            else if (IsChannelType(NonAsyncReturnType, out var channelItemType))
+            {
+                IsChannel = true;
+                StreamReturnType = channelItemType;
+            }
+        }
+
+        private static bool IsChannelType(Type type, out Type payloadType)
+        {
+            var channelType = type.AllBaseTypes().FirstOrDefault(t => t.IsGenericType && t.GetGenericTypeDefinition() == typeof(ChannelReader<>));
+            if (channelType == null)
+            {
+                payloadType = null;
+                return false;
+            }
+
+            payloadType = channelType.GetGenericArguments()[0];
+            return true;
+        }
+
+        private static bool IsObservableType(Type type, out Type payloadType)
+        {
+            var observableInterface = IsIObservable(type) ? type : type.GetInterfaces().FirstOrDefault(IsIObservable);
+            if (observableInterface == null)
+            {
+                payloadType = null;
+                return false;
+            }
+
+            payloadType = observableInterface.GetGenericArguments()[0];
+            return true;
+
+            bool IsIObservable(Type iface)
+            {
+                return iface.IsGenericType && iface.GetGenericTypeDefinition() == typeof(IObservable<>);
+            }
+        }
+
+        public IAsyncEnumerator<object> FromObservable(object observable, CancellationToken cancellationToken)
+        {
+            if (_convertToEnumerator == null)
+            {
+                _convertToEnumerator = CreateConversionMethod(_fromObservableMethod.MakeGenericMethod(StreamReturnType));
+            }
+
+            return _convertToEnumerator.Invoke(observable, cancellationToken);
+        }
+
+        public IAsyncEnumerator<object> FromChannel(object channel, CancellationToken cancellationToken)
+        {
+            if (_convertToEnumerator == null)
+            {
+                _convertToEnumerator = CreateConversionMethod(_getAsyncEnumeratorMethod.MakeGenericMethod(StreamReturnType));
+            }
+
+            return _convertToEnumerator.Invoke(channel, cancellationToken);
+        }
+
+        private Func<object, CancellationToken, IAsyncEnumerator<object>> CreateConversionMethod(MethodInfo methodInfo)
```

Comment:
Add some comments to this method detailing the final codegen. Something like this https://github.com/aspnet/HttpAbstractions/blob/9119433ab778d7adf04c4601917809f2f4582a93/src/Microsoft.AspNetCore.Http.Abstractions/Extensions/UseMiddlewareExtensions.cs#L147-L171